### PR TITLE
Handle Transform features that need access to all values of entity

### PR DESCRIPTION
**Pull request by bschreck**
Adds logic to pass in all values of an entity before cutoff to features that need them.

The problem:
Features like Percentile and CumMean/CumMax/etc. need access to the all the values the entity before the cutoff time, rather than only specified instance ids. Normal Transform features are one-to-one mappings, where the output for specific instance only depends on a single input value for that instance. On the contrary, features like Percentile use all the available data to determine the rank of a particular value. Currently, they are only provided with a subset of the data- the subset that is linked to the set of specified instance ids on target entity.

This PR creates an additional `entity_frames` object inside of `pandas_backend.py`, which contains all the data before the cutoff time. It adds logic to determine whether to pass in this object of the normal `entity_frames` object to each feature calculation function. It also decides where to place the output of these functions.

Lastly, it makes sure to only extract the relevant columns from the entityset using `EntitySet.get_pandas_data_slice`. Previously, we simply used all the columns (which is a waste of memory).

### CODE REVIEW

**Comment by Seth-Rothschild** [comment]
Trying to use this with custom primitives now: what's the expected workflow? I thought that you might be able to set `use_previous` by doing
```
def primitive1(value):
    numeric = pd.Series(value)
    return (numeric - numeric.shift(1))

Prim1 = make_trans_primitive(primitive1,
                            input_types=[vtypes.Numeric],
                            cls_attributes={'use_previous': True},
                            return_type=vtypes.Numeric)
```
but that doesn't seem to be the way to access calculating using additional data.

**Comment by bschreck** [comment]
@Seth-Rothschild the way to do it would be `cls_attributes={'needs_all_values': True}`. You just have the wrong variable name.

File: featuretools/computational_backends/feature_tree.py [hunk 1]
```diff
@@ -25,16 +25,21 @@ def __init__(self, entityset, features, ignored=None):
...
             deps = f.get_deep_dependencies(ignored=ignored)
             feature_deps[f.hash()] = deps
             for dep in deps:
                 all_features[dep.hash()] = dep
-                feature_deps[dep.hash()] = dep.get_deep_dependencies(ignored=ignored)
+                feature_deps[dep.hash()] = dep.get_deep_dependencies(
+                    ignored=ignored)
         self.all_features = list(all_features.values())
         self.feature_deps = feature_deps
+        self.feature_dependents = self._build_dependents()
```
Comment (kmax12) [inline featuretools/computational_backends/feature_tree.py]: i think this is the inverse of `self.feature_deps`, but the variable name doesn't make it very clear since it is the same

Comment (kmax12) [inline featuretools/computational_backends/feature_tree.py]: why break this out into a different function? can it be combined into the logic above that calculates `self.feature_deps`?

File: featuretools/computational_backends/pandas_backend.py [hunk 2]
```diff
@@ -151,15 +169,30 @@ def calculate_all_features(self, instance_ids, time_last,
                 if not self.entityset.find_backward_path(start_entity_id=filter_eid,
                                                          goal_entity_id=eid):
                     entity_frames[eid] = eframes_by_filter[eid][eid]
+                    if (large_eframes_by_filter is not None and
+                            eid in large_eframes_by_filter and eid in large_eframes_by_filter[eid]):
+                        large_entity_frames[eid] = large_eframes_by_filter[eid][eid]
```
Comment (kmax12) [inline featuretools/computational_backends/pandas_backend.py]: what is this line about? seems a bit weird

File: featuretools/computational_backends/pandas_backend.py [hunk 3]
```diff
@@ -151,15 +169,30 @@ def calculate_all_features(self, instance_ids, time_last,
...
                     if verbose:
 
                         pbar.set_postfix({'running': 0})
 
-                    handler = self._feature_type_handler(group[0])
-                    handler(group, entity_frames)
+                    test_feature = group[0]
+
+                    input_frames, output_frames = \
+                        self._determine_input_output_frames(
```
Comment (kmax12) [inline featuretools/computational_backends/pandas_backend.py]: seems like this logic can be done in a simpler way to avoid implementing a function that only gets called once. 

```
input_frames = large_entity_frames
if self.feature_tree.needs_all_values_differentiator(test_feature) == 'normal_no_dependent':
    input_frames = entity_frames
```

then you can put the output frame logic in `existing _place_in_output_frames` function you've defined.

File: featuretools/computational_backends/pandas_backend.py [hunk 4]
```diff
@@ -210,6 +243,72 @@ def generate_default_df(self, instance_ids, extra_columns=None):
                     default_df[c] = [np.nan] * len(instance_ids)
         return default_df
 
+    def _find_necessary_columns(self, only_if_needs_all_values=False):
```
Comment (kmax12) [inline featuretools/computational_backends/pandas_backend.py]: is `only_if_needs_all_values` an optimization or is it required for this to work? a bit hard to follow its role. in general, I find it confusing when a function has a single boolean parameter that controls behavior like this unless it is very obvious what it does. as far as I can tell, it there is only one place in the code where we set it to true. if that is the case, I'd prefer if the optimization happens more explicitly outside the function. makes the function itself more readable and maintainable.

File: featuretools/computational_backends/feature_tree.py (hunk 1 above)
Comment (bschreck) [reply featuretools/computational_backends/feature_tree.py]: we would have to change the above logic to accommodate it, since we rely on get_deep_dependencies() but we don't have an equivalent get_deep_dependents() method. It would just have to be a bit more complicated as a queue/search/while loop instead of just the two for loops. But yeah might make it look more uniform

Comment (bschreck) [reply featuretools/computational_backends/feature_tree.py]: i didn't want to change feature_deps when I was first doing it for fear it was used elsewhere, but now that everythign works it would be better to name it feature_dependencies

File: featuretools/entityset/entityset.py [hunk 5]
```diff
@@ -256,6 +272,7 @@ def query_entity_by_values(self, entity_id, instance_vals, variable_id=None,
                                       variable_id=variable_id,
                                       columns=columns,
                                       time_last=time_last,
+                                      training_window=training_window,
```
[the remark this answers is not in the document]
Comment (bschreck) [reply featuretools/entityset/entityset.py]: I think they weren't

[the remark this answers is not in the document]
Comment (bschreck) [reply featuretools/entityset/entityset.py]: should write a test to confirm

File: featuretools/computational_backends/feature_tree.py (hunk 1 above)
Comment (bschreck) [reply featuretools/computational_backends/feature_tree.py]: Nevermind i think i can make it work

File: featuretools/entityset/entityset.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (bschreck) [reply featuretools/entityset/entityset.py]: We actually never use this function except in tests... We always call the one on the Entity itself (`Entity.query_by_values`). I think we should just delete this function and change those tests to the one we actually use in EntitySet

[the remark this answers is not in the document]
Comment (bschreck) [reply featuretools/entityset/entityset.py]: Changed in latest commit. Let me know what you think

[the remark this answers is not in the document]
Comment (kmax12) [reply featuretools/entityset/entityset.py]: yep, let's remove if it's unused

**Comment by rwedge** [comment]
Codecov has `feature_tree.get_all_features` not being hit by the test cases and it doesn't look like it's called by other parts of featuretools.  Is this a method we want to keep?

File: featuretools/computational_backends/pandas_backend.py [hunk 6]
```diff
@@ -151,15 +171,48 @@ def calculate_all_features(self, instance_ids, time_last,
...
+
+                    input_frames = large_entity_frames
+                    if needs_all_values_type == "normal_no_dependent":
+                        input_frames = entity_frames
+
+                    handler = self._feature_type_handler(test_feature)
+                    result_frame = handler(group, input_frames)
+
+                    output_frames = []
+                    if needs_all_values_type.startswith('dependent'):
```
Comment (kmax12) [inline featuretools/computational_backends/pandas_backend.py]: I'd make this more explicit like the other conditionals around it

```
if needs_all_values_type in ["dependent", "dependent_and_output"]:
    pass
```

File: featuretools/computational_backends/feature_tree.py [hunk 7]
```diff
@@ -25,21 +27,62 @@ def __init__(self, entityset, features, ignored=None):
...
-                feature_deps[dep.hash()] = dep.get_deep_dependencies(ignored=ignored)
+                subdeps = dep.get_deep_dependencies(
+                    ignored=ignored)
+                feature_dependencies[dep.hash()] = subdeps
+                for sd in subdeps:
+                    feature_dependents[sd.hash()].add(dep.hash())
+        # turn values which were hashes of features into the features themselves
+        # (note that they were hashes to prevent checking equality of feature objects,
+        #  which is not currently an allowed operation)
+        self.feature_dependents = {fhash: [all_features[dhash] for dhash in feature_dependents[fhash]]
```
Comment (kmax12) [inline featuretools/computational_backends/feature_tree.py]: why do you iterate over the keys in `all_features` rather than `feature_dependents`?

Comment (bschreck) [reply featuretools/computational_backends/feature_tree.py]: If a feature has no dependencies, it won't be put in feature_dependents by the above code. But, looking at it now, I can just explicitly add those in the for loop.

Comment (bschreck) [reply featuretools/computational_backends/feature_tree.py]: hmm actually that doesn't make much sense because it's a defaultdict, so a feature with no deps will get treated as an empty value in the defaultdict anyway.

File: featuretools/computational_backends/pandas_backend.py [hunk 8]
```diff
@@ -95,13 +97,28 @@ def calculate_all_features(self, instance_ids, time_last,
...
                                                  index_eid=self.target_eid,
                                                  instances=instance_ids,
+                                                 entity_columns=necessary_columns,
                                                  time_last=time_last,
                                                  training_window=training_window,
                                                  verbose=verbose)
+        large_eframes_by_filter = None
+        if any([f.needs_all_values for f in self.feature_tree.all_features]):
+            large_necessary_columns = self.feature_tree.necessary_columns_for_all_values_features
+            all_instances = self.entityset[self.target_eid].get_all_instances()
```
[the remark this answers is not in the document]
Comment (bschreck) [reply featuretools/computational_backends/pandas_backend.py]: workaround for issue #93. can merge with master and remove

File: featuretools/computational_backends/feature_tree.py [hunk 9]
```diff
@@ -153,9 +197,44 @@ def _get_feature_depths(self, entity_id):
...
+                return True
+        return False
+
+    def _is_output_feature(self, f):
+        return f.hash() in self.feature_hashes
 
 # These functions are used for sorting and grouping features
 
+    def needs_all_values_differentiator(self, f):
+        if self._dependent_needs_all_values(f) and self._is_output_feature(f):
```
[the remark this answers is not in the document]
Comment (bschreck) [reply featuretools/computational_backends/feature_tree.py]: yeah i agree

[the remark this answers is not in the document]
Comment (bschreck) [reply featuretools/computational_backends/feature_tree.py]: i think they're more to help me think while I go about solving the problem, and makes sense to review them at final solution

File: featuretools/primitives/cum_transform_feature.py [hunk 10]
```diff
@@ -19,6 +19,7 @@ class CumFeature(TransformPrimitive):
     allow_where = True
     agg_feature = None
     rolling_function = True
+    needs_all_values = True
```
[the remark this answers is not in the document]
Comment (bschreck) [reply featuretools/primitives/cum_transform_feature.py]: What about something like "cross_entity", or "spans_entity"

[the remark this answers is not in the document]
Comment (bschreck) [reply featuretools/primitives/cum_transform_feature.py]: "full_entity_transform"

File: featuretools/computational_backends/feature_tree.py [hunk 11]
```diff
@@ -153,9 +198,53 @@ def _get_feature_depths(self, entity_id):
...
+        is_output = f.hash() in self.feature_hashes
+        # If a dependent feature requires all the instance values
+        # of the associated entity, then we need to calculate this
+        # feature on all values
+        # If the feature is one in which the user requested as
+        # an output (meaning it's part of the input feature list
+        # to calculate_feature_matrix), then we also need
+        # to subselect the output based on the desired instance ids
+        # and place in the return data frame.
+        if self._dependent_needs_all_values(f) and is_output:
```
Comment (kmax12) [inline featuretools/computational_backends/feature_tree.py]: i think you can refactor this to get rid of extra functions

```
dependent_needs_all_values = False
for d in self.feature_dependents[f.hash()]:
        if d.needs_all_values:
            dependent_needs_all_values = True
needs_all_values = f.needs_all_values or dependent_needs_all_values
```
then use variables below

File: featuretools/computational_backends/pandas_backend.py [hunk 12]
```diff
@@ -10,9 +10,9 @@
 
 import numpy as np
 import pandas as pd
+# progress bar
```
Comment (kmax12) [inline featuretools/computational_backends/pandas_backend.py]: delete?

File: featuretools/computational_backends/pandas_backend.py [hunk 13]
```diff
@@ -25,9 +25,11 @@
     IdentityFeature,
     TransformPrimitive
 )
-# progress bar
 from featuretools.utils.gen_utils import make_tqdm_iterator
 
+# featuretools
```
Comment (kmax12) [inline featuretools/computational_backends/pandas_backend.py]: delete?

File: featuretools/computational_backends/pandas_backend.py [hunk 14]
```diff
@@ -151,15 +170,48 @@ def calculate_all_features(self, instance_ids, time_last,
                 if not self.entityset.find_backward_path(start_entity_id=filter_eid,
                                                          goal_entity_id=eid):
                     entity_frames[eid] = eframes_by_filter[eid][eid]
+                    # precalculated features will only be placed in entity_frames,
```
Comment (kmax12) [inline featuretools/computational_backends/pandas_backend.py]: is it possible to have to have precalculated a `needs_all_values` feature? do we need to add code to handle that case?

File: featuretools/computational_backends/pandas_backend.py [hunk 15]
```diff
@@ -151,15 +170,48 @@ def calculate_all_features(self, instance_ids, time_last,
...
+                    if needs_all_values_type != "dependent":
+                        # input was selected (small) set of instances, or
+                        # feature itself is an output,
+                        # so need to place result
+                        # in selected (small) output frame
+                        output_frames.append(entity_frames)
+
+                    for frames in output_frames:
+                        index = frames[entity_id].index
+                        frames[entity_id] = result_frame.loc[index]
```
Comment (kmax12) [inline featuretools/computational_backends/pandas_backend.py]: you might want to use `result_frame.reindex()` here. you want to if some of the index values are going to be in result_frame.

File: featuretools/primitives/cum_transform_feature.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (kmax12) [reply featuretools/primitives/cum_transform_feature.py]: `historical_transform`

[the remark this answers is not in the document]
Comment (kmax12) [reply featuretools/primitives/cum_transform_feature.py]: I like the idea of `full_entity_transform`. any thoughts on other versions that don't repeat transform?

[the remark this answers is not in the document]
Comment (bschreck) [reply featuretools/primitives/cum_transform_feature.py]:
```
applied_to_full_entity
applied_to_full_data
```?

File: featuretools/computational_backends/feature_tree.py [hunk 16]
```diff
@@ -25,28 +27,70 @@ def __init__(self, entityset, features, ignored=None):
...
+        feature_dependents = defaultdict(set)
         for f in features:
             deps = f.get_deep_dependencies(ignored=ignored)
-            feature_deps[f.hash()] = deps
+            feature_dependencies[f.hash()] = deps
             for dep in deps:
+                feature_dependents[dep.hash()].add(f.hash())
                 all_features[dep.hash()] = dep
-                feature_deps[dep.hash()] = dep.get_deep_dependencies(ignored=ignored)
+                subdeps = dep.get_deep_dependencies(
```
[the remark this answers is not in the document]
Comment (bschreck) [reply featuretools/computational_backends/feature_tree.py]: get_deep_dependencies() gets all the dependencies recursively, so features + each one's deep dependencies defines every feature. To build the dependents dict, for each one of those, we need all of its dependents, not dependencies. deps is every dependency of the top level features, and subdeps is every dependency of all the other ones. So deps also contains all the dependents of all the subdeps, and top level features are all the other dependents for all the rest. kind of tricky to wrap your head around

File: featuretools/computational_backends/pandas_backend.py (hunk 14 above)
Comment (bschreck) [reply featuretools/computational_backends/pandas_backend.py]: i thought that I considered all cases before but I did miss the case when you have a `needs_all_values` feature that depends on an approximated/precalculated feature. This will need to be reworked.

**Comment by kmax12** [comment]
Looks great. Thanks for the work on this!
